Исправление логики пригласительных сигналов

angelus1637 opened the pull request:
В текущей версии логика неверно назначает пригласительный сигнал на белую линзу. 
Если в светофоре указаны линзы "BWR", self.GoodInvationSignal = 2, а белая линза будет являться пригласительным сигналом и загорится по команде !sopps. Это неверно.

В рамках исправленной логики, self.GoodInvationSignal всегда равен 0 и изменит своё значение на индекс белой линзы, только в том случае, если она находится под красной линзой, что является необходимым и достаточным условием.
Так, для светофора линзами "WYG-RW", self.GoodInvationSignal = 5, и нижняя белая линза (из двух) будет являться ПСом. А в случае с "BWR", self.GoodInvationSignal = 0. 

Соответствующая проверка добавлена в команды !sopps и !sclps, на self.GoodInvationSignal > 0, чтобы ПС открывался/закрывался только на тех сигналах, где он действительно есть.

angelus1637 committed:
Исправление логики пригласительных сигналов

angelus1637 committed:
Исправление опечаток

alexandr-sizov commented:
Абсолютно бредовая логика, с проверкой "под красным".
 ПС вполне себе может быть не только под красным
Типичная разлинзовка BYR-YW или YR-YW

angelus1637 commented:
> Абсолютно бредовая логика, с проверкой "под красным".
> ПС вполне себе может быть не только под красным
> Типичная разлинзовка BYR-YW или YR-YW

Смысл логики не в том, чтобы ПС был СРАЗУ ПОСЛЕ КРАСНОГО, а чтобы он не был ВЫШЕ ЧЕМ КРАСНЫЙ

angelus1637 commented:
Я обнаружил проблему случайно, когда смог открыть ПС на светофоре Е, который стоит в лоб движению. ПСом он оборудован быть не может, но имеет в себе одну белую линзу и одну красную (WR). 
Нынешняя логика в рамках разлинзовки WR или BWR (при 2/6) находит на светофоре белую линзу, и назначает на неё ПС, но белая линза там всего одна, и она не для ПС. В этом и есть ошибка. 
В случае, когда белых линз две - YWY-RW или WYG-YRW - всё работает исправно, и ПС назначается на последнюю из двух (т.е. на самую нижнюю, которая находится под красной). 
Таким образом, возникла потребность пофиксить назначение ПС на любую белую линзу, которая стоит ВЫШЕ ЧЕМ КРАСНАЯ ЛИНЗА.

Alexell commented:
> Абсолютно бредовая логика, с проверкой "под красным". ПС вполне себе может быть не только под красным Типичная разлинзовка BYR-YW или YR-YW

Алекесандр, вы просто не так поняли. Приведенные вами примеры должны и будут работать при этой логике. А вот такие как WR или BWR не будут более открывать ПС.

alexandr-sizov commented:
Принимается. Код тогда не почитал.

alexandr-sizov commented:
кстати вот вам вариация, КБК, ну допустим, редкая очень, но в принципе имеет место быть.

angelus1637 commented:
> кстати вот вам вариация, КБК, ну допустим, редкая очень, но в принципе имеет место быть.

Хм, тогда ПС назначится, хотя не должен.. Окееей…
Я подумаю, как это можно обработать

angelus1637 committed:
Новая логика назначения ПС

Цикл по-прежнему перебирает все линзы подряд, но теперь проверяет - является ли белой самая нижняя линза в каждой светофорной головке. Если да, то её индекс записывается. Если таких будет несколько, то ПС будет назначен на самую нижнюю из всех.
Варианты разлинзовки:
1. "YYG-RW" - будут записаны индексы 3 и 5, а ПС назначен на 5 линзу.
2. "WYW-RW-W-M" - будут записаны индексы 3, 5 и 6, а ПС назначен на 6 линзу.
3. "WRW" - будет записан индекс 3, а ПС назначен на 3 линзу.
3. "RWR-WR", "WR", "BWR" - ни одна белая линза не подходит под условия, т.к. не является нижней в своей светофорной головке. ПС не будет назначен и self.GoodInvationSignal = 0

angelus1637 commented:
Поправочка, при "YYG-RW" - будет записан индекс 5, и ПС назначен на 5 линзу.

angelus1637 committed:
Финальная версия

Очищен код, убраны все лишние переменные. Проверено на всех типах светофоров. Ошибка назначения ПС исключается на 99%. Единственный случай, когда логика отработает неверно - YW-GR - и подобные, т.е. когда в нижней секции сигнала нет белого цвета, а в верхняя заканчивается белым, однако, это нарушение принципов расположения линз в сигнале.
